relations/src/relations/update_note.rs

```diff
@@ -0,0 +1,157 @@
+};
+
+pub struct UpdateNoteInput<F, A, const MAX_PATH_LEN: usize>
+where
+    F: BigPrimeField,
+    A: CircuitAccount<F>,
+{
+    //public inputs
+    pub op_pub: <A::Op as CircuitOperation<F>>::OpPub,
+    pub new_note_hash: AssignedValue<F>,
+    pub merkle_root: AssignedValue<F>,
+    pub old_nullifier: AssignedValue<F>,
+
+    //witnesses
+    pub new_note: CircuitNote<F>,
+    pub old_note: CircuitNote<F>,
+    pub new_trapdoor: AssignedValue<F>,
+    pub old_trapdoor: AssignedValue<F>,
+    pub new_nullifier: AssignedValue<F>,
+    pub merkle_proof: CircuitMerkleProof<F, MAX_PATH_LEN>,
+    pub op_priv: <A::Op as CircuitOperation<F>>::OpPriv,
+    pub note_id: AssignedValue<F>,
+
+    pub old_account: A,
+}
+
+//helper functions
+#[allow(clippy::too_many_arguments)]
+impl<F, A, const MAX_PATH_LEN: usize> UpdateNoteInput<F, A, MAX_PATH_LEN>
+where
+    F: BigPrimeField,
+    A: CircuitAccount<F>,
+{
+    pub fn new(
+        op_pub: <A::Op as CircuitOperation<F>>::OpPub,
+        new_note_hash: AssignedValue<F>,
+        merkle_root: AssignedValue<F>,
+        old_nullifier: AssignedValue<F>,
+        new_note: CircuitNote<F>,
+        old_note: CircuitNote<F>,
+        new_trapdoor: AssignedValue<F>,
+        old_trapdoor: AssignedValue<F>,
+        new_nullifier: AssignedValue<F>,
+        merkle_proof: CircuitMerkleProof<F, MAX_PATH_LEN>,
+        op_priv: <A::Op as CircuitOperation<F>>::OpPriv,
+        note_id: AssignedValue<F>,
+        old_account: A,
+    ) -> Self {
+        Self {
+            op_pub,
+            new_note_hash,
+            merkle_root,
+            old_nullifier,
+            new_note,
+            old_note,
+            new_trapdoor,
+            old_trapdoor,
+            new_nullifier,
+            merkle_proof,
+            op_priv,
+            note_id,
+            old_account,
+        }
+    }
+}
+
+fn verify_note_circuit<F>(
+    ctx: &mut Context<F>,
+    gate: &GateChip<F>,
+    poseidon: &mut PoseidonHasher<F, T_WIDTH, RATE>,
+    note: &CircuitNote<F>,
+    note_hash: AssignedValue<F>,
+) where
+    F: BigPrimeField,
+{
+    let inner_note_hash = poseidon.hash_fix_len_array(ctx, gate, &note.to_array());
+    let eq = gate.is_equal(ctx, note_hash, inner_note_hash);
+    gate.assert_is_const(ctx, &eq, &F::ONE);
+}
+
+#[allow(dead_code)]
+pub fn update_note_circuit<F, A, const MAX_PATH_LEN: usize>(
+    ctx: &mut Context<F>,
+    input: UpdateNoteInput<F, A, MAX_PATH_LEN>,
+    make_public: &mut Vec<AssignedValue<F>>,
+) where
+    F: BigPrimeField,
+    A: CircuitAccount<F>,
+{
+    let op_pub = input.op_pub;
+
+    make_public.extend(op_pub.clone().into());
+
+    let outer_new_note_hash = input.new_note_hash;
+    let merkle_root = input.merkle_root;
+    let old_nullifier = input.old_nullifier;
+
+    make_public.extend([outer_new_note_hash, merkle_root, old_nullifier]);
+
+    let _new_trapdoor = input.new_trapdoor;
+    let _new_nullifier = input.new_nullifier;
```

Comment:
Added a comment. I was planning to add a test for it

Reply:
Please avoid TODOs in code. If something needs to be done, make it a jira task, apart from that write a comment on why exactly is it written like that because the next person who encounters this will be confused.